### ISSUE TITLE
How to get metrics (auc/accuracy/loss/precision/recall/f1) for both training/validation/test

### ISSUE THREAD

**Issue by rajesh-iiith** [facebookresearch/dlrm#204]
I am using --test-freq=1024 and --mlperf-logging. This seems to be printing all these stats after every 1024th iteration. Are these stats collected on test data? Where can I get similar stats per nth iteration for train and validation data as well?

**Comment by mnaumovfb** [comment]
Please take a look at the code [here](https://github.com/facebookresearch/dlrm/blob/main/dlrm_s_pytorch.py#L1604). The `--print-freq` will print stats on the training data and `--test-freq` on test data. In Criteo datasets the last day is broken into test and validation data and the latter is not used.

**Comment by rajesh-iiith** [comment]
On training data, just loss is being printed, how can I print other stats as well? On --test-freq however I am able to print other stats using --ml-perf logging, but it doesn't print loss for the testing step.

In the paper, I can see that you have plotted validation and training accuracy as well? How did you get the validation numbers?

**Comment by mnaumovfb** [comment]
Loss can be computed per batch/iteration, while other metrics are computed for the entire data set. Therefore, it only makes sense to compute the latter for the test set (if you wish you can add the code to compute them for the entire training set at the end of training). 

We used to compute the accuracy, but later removed for efficiency reasons. You can still some of the computation code for it [commented out](https://github.com/facebookresearch/dlrm/blob/main/dlrm_s_pytorch.py#L1578). You can go back a year in commits and see how it was done and reenable it if needed.

**Comment by rajesh-iiith** [comment]
> On training data, just loss is being printed, how can I print other stats as well? On --test-freq however I am able to print other stats using --ml-perf logging, but it doesn't print loss for the testing step.
> 
> A snippet from the output is below.
> _Finished training it 297984/306969 of epoch 0, 7.01 ms/it, loss 0.508951 Finished training it 299008/306969 of epoch 0, 7.37 ms/it, loss 0.509119 Finished training it 300032/306969 of epoch 0, 7.74 ms/it, loss 0.508069 Finished training it 301056/306969 of epoch 0, 7.92 ms/it, loss 0.505126 Finished training it 302080/306969 of epoch 0, 7.96 ms/it, loss 0.506132 Finished training it 303104/306969 of epoch 0, 7.87 ms/it, loss 0.507316 Finished training it 304128/306969 of epoch 0, 6.98 ms/it, loss 0.506930 Finished training it 305152/306969 of epoch 0, 6.68 ms/it, loss 0.505106 Finished training it 306176/306969 of epoch 0, 6.85 ms/it, loss 0.507457 Finished training it 306969/306969 of epoch 0, 6.79 ms/it, loss 0.507747 :::MLLOG {"namespace": "", "time_ms": 1630256207701, "event_type": "INTERVAL_START", "key": "eval_start", "value": null, "metadata": {"file": "dlrm_s_pytorch.py", "lineno": 1726, "epoch_num": 2.0}} Testing at - 306969/306969 of epoch 0, recall 0.2288, precision 0.6008, f1 0.3314, ap 0.4921, auc 0.7225, best auc 0.7225, accuracy 75.870 %, best accuracy 0.000 % Saving model to pt_models/dlrm_model_45M_1epoch_1.pt_
> 
> Also, In the paper, I can see that you have plotted validation and training accuracy as well? How did you get the validation numbers?

In this log, we see **testing at x/y of epoch n**. This takes place on the validation set or the test set?

**Comment by mnaumovfb** [comment]
It takes place on the test set. The validation set is not used.
